main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer1.scala

```diff
@@ -0,0 +1,580 @@
+    }
+
+    def get(varSym: VarSym): Occur = {
+      this.vars.getOrElse(varSym, Dead)
+    }
+  }
+
+  private def increment(occurInfo: OccurInfo): OccurInfo = {
+    occurInfo.copy(size = occurInfo.size + 1)
+  }
+
+  /**
+    * Performs occurrence analysis on the given AST `root`.
+    */
+  def run(root: MonoAst.Root)(implicit flix: Flix): OccurrenceAst1.Root = {
+    val defs = visitDefs(root.defs) // visitDefs call parMap internally and has additional handling
+    val effects = ParOps.parMapValues(root.effects)(visitEffect)
+    val enums = ParOps.parMapValues(root.enums)(visitEnum)
+    val structs = ParOps.parMapValues(root.structs)(visitStruct)
+    OccurrenceAst1.Root(defs, enums, structs, effects, root.mainEntryPoint, root.entryPoints, root.sources)
+  }
+
+  /**
+    * Performs occurrence analysis on every entry in `defs0` in parallel.
+    * Decorates each Def with occurrence information, i.e., how it appears in the program or if it is unused.
+    */
+  private def visitDefs(defs0: Map[DefnSym, MonoAst.Def])(implicit flix: Flix): Map[DefnSym, OccurrenceAst1.Def] = {
+    val (ds, os) = ParOps.parMap(defs0.values)(visitDef).unzip
+
+    // Combine all `defOccurrences` into one map.
+    val defOccur = combineAll(os)
+
+    // Updates the occurrence of every `def` in `ds` based on the occurrence found in `defOccur`.
+    ds.foldLeft(Map.empty[DefnSym, OccurrenceAst1.Def]) {
+      case (macc, defn) =>
+        val occur = defOccur.getOrElse(defn.sym, Dead)
+        val newContext = defn.context.copy(occur = occur)
+        val defWithContext = defn.copy(context = newContext)
+        macc + (defn.sym -> defWithContext)
+    }
+  }
+
+  /**
+    * Performs occurrence analysis on `defn`.
+    */
+  private def visitDef(defn0: MonoAst.Def): (OccurrenceAst1.Def, OccurInfo) = {
+
+    /**
+      * Returns true if `expr0` is a function call.
+      */
+    def isDirectCall(expr0: OccurrenceAst1.Expr): Boolean = expr0 match {
+      case OccurrenceAst1.Expr.ApplyDef(_sym, _exps, _, _, _, _) => true
+      // sym != defn0.sym // && exps.forall(isTrivial) // see doc comment for DefContext
+
+      case OccurrenceAst1.Expr.ApplyClo(clo, _exps, _, _, _) =>
+        clo match {
+          case OccurrenceAst1.Expr.ApplyAtomic(AtomicOp.Closure(_sym), _, _, _, _) => true
+          // sym != defn0.sym // && exps.forall(isTrivial) // see doc comment for DefContext
+          case _ => false
+        }
+      case _ => false
+    }
+
+    /**
+      * Returns true if `def0` occurs in `occurInfo` without being captured.
+      */
+    def isSelfRecursive(occurInfo: OccurInfo): Boolean = occurInfo.defs.get(defn0.sym) match {
+      case None => false
+      case Some(o) => o match {
+        case Occur.Dead => false
+        case Occur.Once => true
+        case Occur.OnceInAbstraction => true
+        case Occur.Many => true
+        case Occur.ManyBranch => true
+        case Occur.DontInline => false
+        case Occur.Dangerous => true
```

Comment:
It feels like there is some information loss here - couldn't self-recursive be its on field on OccurInfo?